proof/verifier.go

```diff
@@ -165,9 +176,71 @@ func verifyTaprootProof(anchor *wire.MsgTx, proof *TaprootProof,
 
 // verifyInclusionProof verifies the InclusionProof is valid.
 func (p *Proof) verifyInclusionProof() (*commitment.TapCommitment, error) {
-	return verifyTaprootProof(
+	// We always check the v0 inclusion proofs.
+	v0Commitment, err := verifyTaprootProof(
 		&p.AnchorTx, &p.InclusionProof, &p.Asset, true,
 	)
+	if err != nil {
+		return nil, fmt.Errorf("error verifying v0 inclusion proof: %w",
+			err)
+	}
+
+	// If this is a v1 proof, we decide whether we need to have a v1
+	// inclusion proof based on whether this is a transfer root asset.
+	needStxoProofs := p.IsVersionV1() && p.Asset.IsTransferRoot()
+	hasStxoProofs := p.InclusionProof.CommitmentProof != nil &&
+		len(p.InclusionProof.CommitmentProof.STXOProofs) > 0
+
+	if needStxoProofs && !hasStxoProofs {
+		return nil, fmt.Errorf("%w: missing STXO proofs",
+			ErrInvalidCommitmentProof)
+	}
+
+	// We ignore the STXO proofs if the proof signals version 0, or if there
+	// are no STXO proofs present (because they're not needed for this type
+	// of asset).
+	if p.IsVersionV0() || !hasStxoProofs {
+		return v0Commitment, nil
+	}
+
+	// For an inclusion proof, there is only one output index that we need
+	// to check.
+	outIdx := p.InclusionProof.OutputIndex
+	p2trOutputs := map[uint32]fn.Set[asset.SerializedKey]{
```

Comment:
You can use the `P2TROutputsSTXOs` type alias defined above to shorten this slightly.